comictalker/comictalker.py

```diff
@@ -0,0 +1,200 @@
+            return comic_data
+        except NotImplementedError:
+            logger.warning(f"{self.talker.source_details.name} has not implemented: 'fetch_comic_data'")
+            raise TalkerError(
+                self.talker.source_details.name,
+                4,
+                "The source has not implemented: 'fetch_comic_data'",
+            )
+
+    # Master function to get issues in a series/volume
+    def fetch_issues_by_volume(self, series_id: int) -> list[ComicIssue]:
+        try:
+            issues_result = self.talker.fetch_issues_by_volume(series_id)
+            return issues_result
+        except NotImplementedError:
+            logger.warning(f"{self.talker.source_details.name} has not implemented: 'fetch_issues_by_volume'")
+            raise TalkerError(
+                self.talker.source_details.name,
+                4,
+                "The source has not implemented: 'fetch_issues_by_volume'",
+            )
+
+    # For CLI
+    def fetch_issue_data_by_issue_id(self, issue_id: int) -> GenericMetadata:
+        try:
+            issue_result = self.talker.fetch_issue_data_by_issue_id(issue_id)
+            return issue_result
+        except NotImplementedError:
+            logger.warning(f"{self.talker.source_details.name} has not implemented: 'fetch_issue_data_by_issue_id'")
+            raise TalkerError(
+                self.talker.source_details.name,
+                4,
+                "The source has not implemented: 'fetch_issue_data_by_issue_id'",
+            )
+
+    # For issueidentifer
+    def fetch_alternate_cover_urls(self, issue_id: int) -> list[str]:
+        try:
+            alt_covers = self.talker.fetch_alternate_cover_urls(issue_id)
+            return alt_covers
+        except NotImplementedError:
+            logger.warning(f"{self.talker.source_details.name} has not implemented: 'fetch_alternate_cover_urls'")
+            raise TalkerError(
+                self.talker.source_details.name,
+                4,
+                "The source has not implemented: 'fetch_alternate_cover_urls'",
+            )
+
+    # For issueidentifier
+    def fetch_issues_by_volume_issue_num_and_year(
+        self, volume_id_list: list[int], issue_number: str, year: str | int | None
+    ) -> list[ComicIssue]:
+        try:
+            issue_results = self.talker.fetch_issues_by_volume_issue_num_and_year(volume_id_list, issue_number, year)
+            return issue_results
+        except NotImplementedError:
+            logger.warning(
+                f"{self.talker.source_details.name} has not implemented: 'fetch_issues_by_volume_issue_num_and_year'"
+            )
+            raise TalkerError(
+                self.talker.source_details.name,
+                4,
+                "The source has not implemented: 'fetch_issues_by_volume_issue_num_and_year'",
+            )
+
+    def fetch_issue_cover_urls(self, issue_id: int) -> tuple[str | None, str | None]:
```

Comment:
This calls `fetch_issue_select_details` which is a stripped down version of `fetch_issue_data_by_issue_id` (or `fetch_issues_by_volume_issue_num_and_year` or `fetch_issues_by_volume`) that returns a dict instead of metadata.
However `fetch_issue_select_details` never actually gets executed because it checks the cache first and the cache is always already filled in. let's remove this and instead just retrieve the whole issue data. We can also remove `class SelectDetails(TypedDict):` from resulttypes.py along with this.

This will make it easier for someone else to implement the next Talker

Reply:
I've mostly done this but am wondering if I have the wrong idea now. The problem is, if I understand correctly, by removing `fetch_issue_cover_urls` and using `fetch_comic_data` (internally to CV `fetch_issue_data_by_issue_id`) that requires a full request to the CV server. This is a problem because `coverimagewidget` uses it. Right now I've done this:
```
    def set_issue_id(self, issue_id: int) -> None:
        if self.mode == CoverImageWidget.AltCoverMode:
            self.reset_widget()
            self.update_content()
            self.issue_id = issue_id

            ComicTalker.url_fetch_complete = self.sig.emit_url
            # self.talker_api.async_fetch_issue_cover_urls(self.issue_id)
            details = self.talker_api.fetch_comic_data(0, "", self.issue_id)
            ComicTalker.url_fetch_complete(details.cover_image, None)
```
So for every issue clicked on, it does a full request to get the image url. The image url is in the cache but `fetch_comic_data` will never touch it.

Of the available functions, the only one that could be used currently is `fetch_issues_by_volume` which would hit the cache but then it's returning a list that will need to be searched for the issue.

Hope that makes sense? Am I missing something obvious?

Reply:
You bring up a good point we should probably cache `fetch_issue_data_by_issue_id `, but we actually don't need it at all (this is why I want to make it simpler, turns out we're doing extra work for no reason), everywhere that `set_issue_id` (from `CoverImageWidget`) is used it already has the appropriate image url or it can be added by adding image to the field_list parameter on the request to CV. We can fully replace `set_issue_id` in each case by using `set_url(self.current_match()["image"])` instead of `set_issue_id(self.current_match()["issue_id"])` or similar code.

Reply:
I think I've covered it all with https://github.com/comictagger/comictagger/pull/341/commits/c9cd58fecb14e72e77b6394cac86d2ebe6a60222

Reply:
Yes that covers all of `fetch_issue_cover_urls`.
Now `fetch_alternate_cover_urls` has an `issue_url` parameter which complicates things because now we are telling that the Talker where to check for alternate covers which isn't helpful, the talker already knows where the alternate covers are if they exist.
So instead of passing the url from the talker and back to it we need to use a function to return a single issue from the source by ID and we should cache it in case it has already been retrieved

Reply:
Yeah, I removed the CV function `fetch_issue_page_url` that was using `fetch_issue_select_details` to check the site url for the issue page. There is no cache function atm that will return a single issue record. As the url was already available (issue_id and now the image_url are being passed in), I thought that was the better option than adding a function to the cacher or using `get_volume_issues_info` and searching through the list it returns. Or, even worse, doing a fresh call to the api. I thought not doing a call to either the cache or api was the better way.

You would still prefer a new cache function? It should always be in the cache I believe but can add an api call as backup anyway.

Reply:
There are two problems:

- The first `issue_url` is not the final URL we need, that URL is the detail page you get when you browse to the page on CV manually, that page is then downloaded and parsed to get the URLs that we do need, and that is specific to CV because their API does not list the alternate covers.

- The second issue is that we are passing `issue_url` back to the `Talker`, which is where it came from, with the `image_url` we are passing it to the `ImageFetcher`, which has no other relation to the `Talker` (though it uses the same cache file which might be a good idea to change). With passing the `issue_url` to the `Talker` we are telling it that it needs to use it in order to get the URLs which is only true for CV, the Marvel official API for example just provides a list of images https://developer.marvel.com/docs#!/public/getComicIndividual_get_7 and therefore the `issue_url` is useless.

Also any cache calls are generally considered free, or at least such a small cost that you should always check it first for literally everything especially, when the data we are using is small, like it is here.

Make a cache function to check the `Issues` table and return that first during `fetch_issue_data_by_issue_id`

Reply:
I understand now that because this is a workaround and another source with alt covers would (you hope) provide their URLs via the API, the workaround needs to be internal to the CV talker.

Sorry if I'm being dense here but `fetch_issue_data_by_issue_id` never checks the cache because the cache will never have the full information required. It stores only a subset of the information needed by the issue list window. So I see a couple of options:
- Have the initial call to populate the issue list window retrieve all the information needed to fulfill the metadata requirements of `fetch_issue_data_by_issue_id` and so the cache then becomes relevant to that call. A new cache function is needed. (I've done a small test to see the difference between a full API issue call and a limited field call and the variance seems to be the same. See below.)
- The other option is to make a cache function to retrieve a single record for the limited issue data and a new function to retrieve only the URL from that cache call. This would be specific to only CV then.

Normally it would make sense to only retrieve the data required to fulfill the issue list window but as it stands there doesn't appear to be a difference in time for the CV API. So the first option makes the most sense. If others will be the same...

Hopefully I've got that all correct and it checks out?

```
Response (0.744s) - https://comicvine.gamespot.com/api/issue/4000-157781/?api_key=apikey&format=json
Response (0.453s) - https://comicvine.gamespot.com/api/issue/4000-157781/?api_key=apikey&format=json
Response (0.28s) - https://comicvine.gamespot.com/api/issue/4000-157781/?api_key=apikey&format=json
Response (0.417s) - https://comicvine.gamespot.com/api/issue/4000-157781/?api_key=apikey&format=json
Response (0.293s) - https://comicvine.gamespot.com/api/issue/4000-157781/?api_key=apikey&format=json
Response (0.277s) - https://comicvine.gamespot.com/api/issue/4000-157781/?api_key=apikey&format=json
```
```
Response (0.5s) - https://comicvine.gamespot.com/api/issue/4000-157781/?api_key=apikey&format=json&field_list=id,volume,issue_number,name,image,cover_date,site_detail_url,description,aliases
Response (0.446s) - https://comicvine.gamespot.com/api/issue/4000-157781/?api_key=apikey&format=json&field_list=id,volume,issue_number,name,image,cover_date,site_detail_url,description,aliases
Response (0.21s) - https://comicvine.gamespot.com/api/issue/4000-157781/?api_key=apikey&format=json&field_list=id,volume,issue_number,name,image,cover_date,site_detail_url,description,aliases
Response (0.199s) - https://comicvine.gamespot.com/api/issue/4000-157781/?api_key=apikey&format=json&field_list=id,volume,issue_number,name,image,cover_date,site_detail_url,description,aliases
Response (0.276s) - https://comicvine.gamespot.com/api/issue/4000-157781/?api_key=apikey&format=json&field_list=id,volume,issue_number,name,image,cover_date,site_detail_url,description,aliases
Response (0.417s) - https://comicvine.gamespot.com/api/issue/4000-157781/?api_key=apikey&format=json&field_list=id,volume,issue_number,name,image,cover_date,site_detail_url,description,aliases
```

Reply:
Yes you are correct on all accounts. (though the 'only retrieve the data required' is arguable, [MongoDB] disagrees with you)

And I think you may have missed my last line here
> Make a cache function to check the `Issues` table and return that first during `fetch_issue_data_by_issue_id`

We should go with your first option, also I believe all of the calls to CV do include the `site_detail_url` which is what we need in this case, though they aren't all cached.

I looked at `ComicCacher` again and noticed that we actually already have a function to cache
```python
add_volume_issues_info(self, source_name: str, volume_id: int, volume_issues: list[ComicIssue])
```
which we can actually just change to 
```python
add_issues_info(self, source_name: str, volume_issues: list[ComicIssue])
```

Because `ComicIssue` has a `volume` key to hold the volume info, and then to call it with a single issue
```python
add_volume_issues_info(self, "comicvine", [issue])
```

And then anywhere that we request issues add a cache call at the end to make sure all the issues end up in the cache (such as `fetch_issues_by_volume_issue_num_and_year`, don't worry about retrieving the cache during this function just cache it's result at the end).
And then a new function should be made `get_issue_info(self, issue_id: int) -> ComicIssue` (or similar) that we use at the beginning of `fetch_issue_data_by_issue_id`

[MongoDB]: https://www.mongodb.com/developer/products/mongodb/mongodb-schema-design-best-practices/

Reply:
`add_volume_issues_info` is the one used for the issue window list. I think I have enough to do it and then you can tell me where I did it wrong :)

Reply:
Okay, I was being stupid. I should have double checked https://comicvine.gamespot.com/api/issues which does not return the same as https://comicvine.gamespot.com/api/issue/xxxx-xxxxxx. I will do a solution but `fetch_issue_data_by_issue_id` cannot use the cached data in any useful way.

Reply:
Okay... how is https://github.com/comictagger/comictagger/pull/341/commits/c5ad75370f3164b20d5f66a271a0c8c02e561cd7 ?

Reply:
This works, but I don't understand what the `issues` endpoint provides that the `issue/xxx-xxxxxx` endpoint doesn't have, it returns the same json but *without* the credit and character information, in other words it has more information than the `issues` endpoint. What's interesting is that CV now actually has an undocumented `associated_images` property which has all of the alt images (they definitely didn't have this the last time I looked).

So merge all the new things added in `fetch_partial_issue_data` to `fetch_issue_data_by_issue_id`.
Also do the same to `fetch_partial_volume_data` and `fetch_volume_data`
Once that is done we can actually remove the entire work around of fetching URLs separately, and use the `associated_images` property from the API.
So `fetch_alternate_cover_urls` and `async_fetch_alternate_cover_urls` can be removed along with `cache_alternate_cover_urls`, `fetch_cached_alternate_cover_urls`, `parse_out_alt_cover_urls`

Reply:
I saw the `associated_images` and meant to look into it and then totally forgot!

I think we're (or more likely me) are having cross-talk about the cache and `fetch_issue_data_by_issue_id` etc. Because `issues` endpoint doesn't have the `persons`, `characters`, etc. that is required for the genericmetadata then a call to the API (`issue/xxxx-xxxxx`) is always required.

The API docs I've found also have other incorrect information about field names. Unless I'm reading it wrong, `volume` says it's `person_credits` but the API returns `people` etc.

I'll see about reducing it all down.

Reply:
Right now there is a separate table for alt image urls. I don't see any reason not to put those into the `issues` table?

Reply:
Yes that should be put into the issues table

Reply:
Take a look at https://github.com/comictagger/comictagger/pull/341/commits/4514ae80d05fa97d1fedeb76180fa35e2f703d6c If I'm being dumb about `fetch_partial_volume_data` let me know :)

Reply:
That's a good point we still need the data, so instead of enforcing a new call to the CV API we are going to extend the database schema. Add columns for characters,people and locations and you can use newlines to separate, if anyone tries to put a newline in the middle of a name I don't care about them or how it ends up looking (people may need it's own table), where as a comma can certainly be in a name.

Reply:
So this is where I'm having trouble because `issues` does not support returning characters, persons etc. Only `issue` which is a single call by id. https://comicvine.gamespot.com/api/documentation#toc-0-10 issues right below. What am I missing? :)

Reply:
1. Cache it, it won't hit the cache very often but it will hit it eventually.
   Once it's cached we have the issue of two different things caching the same `ComicIssue` but one does not have the full data, and we need to ignore the cache when we don't have the needed data.
1. Create a new column named complete that is a boolean, (sqlite is weird and returns an int just use `bool(row[index])` to make it a proper boolean when you retrieve the cache).
1. Now we need a way to tell the cache that what we are putting in is "complete" for this we will need to either:
   1. Add a new parameter that defaults to either true or false on all of the methods that cache a `ComicIssue` or 
   1. Add a "complete" parameter to the `ComicIssue` type.
1. Now that we have the correct data during normal operations everything stays the same as before except we note that it is not complete, then during `fetch_issue_data_by_issue_id` we check if it is complete in addition to whether or not it is in the cache

Reply:
I think it's going to be very rare that the full cache is hit again but it can't hurt. Doing the same for volume should see more use as with manga it's quite usual to have multiple chapters as separate files so using MU, it would be getting the same volume info over and over.

As we're on the subject, this may not be a problem as it could only relate to CV: The `volume` info for people does not have a role. Would using "Unknown" be best for no role if it comes to it?

And when it comes to people, the easiest option instead of doing another table might be to save the person record as JSON; `{"name": "Will Shakesphere", "role": "Writer"}` or maybe something like `Will Shakesphere {Writer}` so the roles can be parsed out?

Reply:
For the volume we don't need to implement it ComicTagger will never use that information as it will use the info from the issue, but the correct role would be artist, the artist role doesn't mean anything in particular it just means they took part in creating it. That's fine just make sure that you add the appropriate changes to the `ComicIssue` and `ComicVolume` classes

Reply:
https://github.com/comictagger/comictagger/pull/341/commits/bf06b94284d5554ee921d58d507ee768e9695a6f should cover all this.